api/envoy/config/route/v3/route_components.proto

```diff
@@ -750,10 +750,49 @@ message RouteAction {
     ConnectConfig connect_config = 3;
   }
 
+  message ClusterSelector {
+    // Indicates the upstream cluster to which the request should be routed
+    // to.
+    string cluster = 1 [(validate.rules).string = {min_bytes: 1}];
+
+    // Envoy will determine the cluster to route to by reading the value of the
+    // HTTP header named by cluster_header from the request headers. If the
+    // header is not found or the referenced cluster does not exist, Envoy will
+    // return a 404 response.
+    //
+    // .. attention::
+    //
+    //   Internally, Envoy always uses the HTTP/2 *:authority* header to represent the HTTP/1
+    //   *Host* header. Thus, if attempting to match on *Host*, match on *:authority* instead.
+    string cluster_header = 2
+        [(validate.rules).string = {min_bytes: 1 well_known_regex: HTTP_HEADER_NAME strict: false}];
+
+    // Multiple upstream clusters can be specified for a given route. The
+    // request is routed to one of the upstream clusters based on weights
+    // assigned to each cluster. See
+    // :ref:`traffic splitting <config_http_conn_man_route_table_traffic_splitting_split>`
+    // for additional documentation.
+    WeightedCluster weighted_clusters = 3;
+
+    // Envoy will determine the cluster to route to by reading the value of the filter state
+    // referenced by the key specified in this field. If the key is not found or the referenced
+    // cluster does not exist, Envoy will return an error response.
+    // NOTE: the `FilterState::Object` pointed to by the key must implement
+    // `serializeAsString()`; the returned value is expected to be the cluster
+    // name.
+    string cluster_filter_state_key = 4 [(validate.rules).string = {min_bytes: 1}];
+  }
+
+  // A wrapper to enable repeated |ClusterSelector|s in a oneof.
+  message ClusterSelectorRepeatedWrapper {
+    repeated ClusterSelector cluster_selectors = 1;
+  }
+
   reserved 12, 18, 19, 16, 22, 21, 10;
 
   reserved "request_mirror_policy";
 
+  // NOTE: This oneof can be deprecated for a `repeated ClusterSelector`.
```

Comment:
If we go with this can we just go ahead and do the deprecation? I agree this is the right path forward.

Reply:
SGTM.

Reply:
+1 on deprecating existing redundant selectors.